src/compute/src/render/reduce.rs

```diff
@@ -630,47 +653,106 @@ where
         move |_key, source, target| {
-            // Negative counts would be surprising, but until we are 100% certain we wont
-            // see them, we should report when we do. We may want to bake even more info
-            // in here in the future.
-            if source.iter().any(|(_val, cnt)| cnt < &0) {
-                // XXX: This reports user data, which we perhaps should not do!
-                for (val, cnt) in source.iter() {
-                    if *cnt < 0 {
-                        warn!("[customer-data] Negative accumulation in ReduceInaccumulable: {val:?} with count {cnt:?} in dataflow {debug_name}");
-                        soft_assert_or_log!(false, "Negative accumulation in ReduceInaccumulable");
+            // We respect the multiplicity here (unlike in hierarchical aggregation)
+            // because we don't know that the aggregation method is not sensitive
+            // to the number of records.
+            let iter = source.iter().flat_map(|(v, w)| {
+                // Note that in the non-positive case, this is wrong, but harmless because
+                // our other reduction will produce an error.
+                let count = usize::try_from(*w).unwrap_or(0);
+                std::iter::repeat(v.iter().next().unwrap()).take(count)
+            });
+            row_buf.packer().push(func.eval(iter, &RowArena::new()));
+            target.push((row_buf.clone(), 1));
+        }
+    });
+
+    if validating && err_output.is_none() {
+        let debug_name = debug_name.to_string();
+        let errs = arranged.reduce_abelian::<_, ErrValSpine<_, _, _>>(
+            "ReduceInaccumulable Error Check",
+            move |_key, source, target| {
+                // Negative counts would be surprising, but until we are 100% certain we won't
+                // see them, we should report when we do. We may want to bake even more info
+                // in here in the future.
+                for (value, count) in source.iter() {
+                    if count.is_positive() {
+                        continue;
                     }
+
+                    let message = "Non-positive accumulation in ReduceInaccumulable";
+                    warn!(?value, ?count, debug_name, "[customer-data] {message}");
+                    error!("{message}");
+                    target.push((
+                        DataflowError::from(EvalError::Internal(message.to_string())),
+                        1,
+                    ));
+                    return;
+                }
+            },
+        );
+        (oks, Some(errs.as_collection(|_, v| v.clone())))
+    } else {
+        (oks, err_output)
+    }
+}
+
+fn build_reduce_inaccumulable_distinct<G, R>(
+    debug_name: &str,
+    input: Collection<G, (Row, Row), Diff>,
+) -> differential_dataflow::operators::arrange::Arranged<
+    G,
+    crate::typedefs::TraceRowHandle<(Row, Row), R, G::Timestamp, Diff>,
+>
+// Arrangement<G, (Row, Row)>
```

Comment:
Should we introduce a `KeyArrangement` type to make this signature fit better with the other functions in this file? Something like:
```Rust
pub(crate) type KeyArrangement<S, K> =
    Arranged<S, TraceRowHandle<K, (), <S as ScopeParent>::Timestamp, Diff>>;
```
I am not sure if the best place for this would be in `reduce.rs` or `context.rs`. Perhaps the latter?

Reply:
Oof, I meant to clean this up.  Creating a type is the right idea, I think, and I was searching for a good name.  I think `KeyArrangement` works.